api/filrewardsd/service/service.go

```diff
@@ -0,0 +1,670 @@
+	}
+	res := &pb.ListRewardsResponse{
+		Rewards: pbRewards,
+		More:    more,
+	}
+	if startAt != nil {
+		res.MoreToken = startAt.UnixNano()
+	}
+	return res, nil
+}
+
+func (s *Service) Claim(ctx context.Context, req *pb.ClaimRequest) (*pb.ClaimResponse, error) {
+	lck := s.semaphores.Get(orgKeyLock(req.OrgKey))
+	lck.Acquire()
+	defer lck.Release()
+
+	totalRewarded, err := s.totalRewarded(ctx, req.OrgKey)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "calculating total rewarded: %v", err)
+	}
+	totalPending, err := s.totalClaimed(ctx, req.OrgKey, pb.ClaimState_CLAIM_STATE_PENDING)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "calculating total claimed: %v", err)
+	}
+	totalClaimed, err := s.totalClaimed(ctx, req.OrgKey, pb.ClaimState_CLAIM_STATE_COMPLETE)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "calculating total claimed: %v", err)
+	}
+
+	available := totalRewarded - totalPending - totalClaimed
+
+	if req.Amount > available {
+		return nil, status.Errorf(codes.InvalidArgument, "claim amount %d is greater than available reward balance %d", req.Amount, available)
+	}
+
+	t := time.Now()
+	c := &claim{
+		ID:        primitive.NewObjectID(),
+		OrgKey:    req.OrgKey,
+		ClaimedBy: req.ClaimedBy,
+		Amount:    req.Amount,
+		State:     pb.ClaimState_CLAIM_STATE_PENDING,
+		CreatedAt: t,
+		UpdatedAt: t,
+	}
+
+	if _, err := s.claimsCol.InsertOne(ctx, c); err != nil {
+		return nil, status.Errorf(codes.Internal, "inserting claim: %v", err)
+	}
+
+	if err := s.ac.Track(
+		ctx,
+		c.OrgKey,
+		analyticspb.AccountType_ACCOUNT_TYPE_ORG,
+		analyticspb.Event_EVENT_FIL_CLAIM,
+		analytics.WithProperties(map[string]interface{}{
+			"id":              c.ID.Hex(),
+			"claimed_by":      c.ClaimedBy,
+			"amount_nano_fil": c.Amount,
+			"amount_fil":      float64(c.Amount) / math.Pow10(9),
+		}),
+	); err != nil {
+		log.Errorf("calling analytics track: %v", err)
+	}
+
+	return &pb.ClaimResponse{Claim: toPbClaim(c)}, nil
+}
+
+func (s *Service) FinalizeClaim(ctx context.Context, req *pb.FinalizeClaimRequest) (*pb.FinalizeClaimResponse, error) {
+	lck := s.semaphores.Get(orgKeyLock(req.OrgKey))
+	lck.Acquire()
+	defer lck.Release()
+
+	var state pb.ClaimState
+	if req.TxnCid != "" && req.FailureMessage == "" {
+		state = pb.ClaimState_CLAIM_STATE_COMPLETE
+	} else if req.FailureMessage != "" && req.TxnCid == "" {
+		state = pb.ClaimState_CLAIM_STATE_FAILED
+	} else {
+		return nil, status.Error(codes.InvalidArgument, "must provide a txn cid or failure message")
+	}
+
+	objID, err := primitive.ObjectIDFromHex(req.Id)
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "parsing claim id: %v", err)
+	}
+
+	res := s.claimsCol.FindOneAndUpdate(
+		ctx,
+		bson.M{"_id": objID},
+		bson.M{"$set": bson.M{"state": state, "txn_cid": req.TxnCid, "failure_message": req.FailureMessage, "updated_at": time.Now()}},
+	)
+	if res.Err() == mongo.ErrNoDocuments {
```

Comment:
Nit: `errors.Is()` is a nice way to compare errors.

Reply:
Updated everywhere.